1- Buscar un conjunto de datos a analizar: 

Se recomienda que el conjunto de datos a analizar tenga variables numéricas y categóricas. Ya que también se analizarán.

In [5]:
import pandas as pd
import numpy as np
from scipy import stats
pd.options.display.max_columns = None


La base de datos que hemos seleccionado es el dataset de Sephora, dejamos el link a mano para poder consultar las descripciones de las columnas
https://www.kaggle.com/datasets/raghadalharbi/all-products-available-on-sephora-website?resource=download


In [6]:
df = pd.read_csv('../datos/sephora_website_dataset.csv')
df.head(2)

id           brand   category                            name  \
0  2218774  Acqua Di Parma  Fragrance  Blu Mediterraneo MINIATURE Set   
1  2044816  Acqua Di Parma    Cologne                         Colonia   

             size  rating  number_of_reviews  love  price  value_price  \
0  5 x 0.16oz/5mL     4.0                  4  3002   66.0         75.0   
1   0.7 oz/ 20 mL     4.5                 76  2700   66.0         66.0   

                                                 URL  MarketingFlags  \
0  https://www.sephora.com/product/blu-mediterran...            True   
1  https://www.sephora.com/product/colonia-P16360...            True   

  MarketingFlags_content                                            options  \
0            online only                                         no options   
1            online only   - 0.7 oz/ 20 mL  Spray  - 1.7 oz/ 50 mL Eau d...   

                                             details  \
0  This enchanting set comes in a specially handc...   
1  An elegant timeless scent filled with a fresh-...   

                                          how_to_use  \
0  Suggested Usage:-Fragrance is intensified by t...   
1                                    no instructions   

                                         ingredients  online_only  exclusive  \
0  Arancia di Capri Eau de Toilette: Alcohol Dena...            1          0   
1                                            unknown            1          0   

   limited_edition  limited_time_offer  
0                0                   0  
1                0                   0

In [7]:
#Exploramos el dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9168 entries, 0 to 9167
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      9168 non-null   int64  
 1   brand                   9168 non-null   object 
 2   category                9168 non-null   object 
 3   name                    9168 non-null   object 
 4   size                    9168 non-null   object 
 5   rating                  9168 non-null   float64
 6   number_of_reviews       9168 non-null   int64  
 7   love                    9168 non-null   int64  
 8   price                   9168 non-null   float64
 9   value_price             9168 non-null   float64
 10  URL                     9168 non-null   object 
 11  MarketingFlags          9168 non-null   bool   
 12  MarketingFlags_content  9168 non-null   object 
 13  options                 9168 non-null   object 
 14  details                 9168 non-null   

In [8]:
columnas_nuevas = {col : col.lower() for col in df.columns} #Cambiamos todas las columnas a minuscula

In [9]:
df.rename(columns = columnas_nuevas, inplace = True)
df.head(2)

id           brand   category                            name  \
0  2218774  Acqua Di Parma  Fragrance  Blu Mediterraneo MINIATURE Set   
1  2044816  Acqua Di Parma    Cologne                         Colonia   

             size  rating  number_of_reviews  love  price  value_price  \
0  5 x 0.16oz/5mL     4.0                  4  3002   66.0         75.0   
1   0.7 oz/ 20 mL     4.5                 76  2700   66.0         66.0   

                                                 url  marketingflags  \
0  https://www.sephora.com/product/blu-mediterran...            True   
1  https://www.sephora.com/product/colonia-P16360...            True   

  marketingflags_content                                            options  \
0            online only                                         no options   
1            online only   - 0.7 oz/ 20 mL  Spray  - 1.7 oz/ 50 mL Eau d...   

                                             details  \
0  This enchanting set comes in a specially handc...   
1  An elegant timeless scent filled with a fresh-...   

                                          how_to_use  \
0  Suggested Usage:-Fragrance is intensified by t...   
1                                    no instructions   

                                         ingredients  online_only  exclusive  \
0  Arancia di Capri Eau de Toilette: Alcohol Dena...            1          0   
1                                            unknown            1          0   

   limited_edition  limited_time_offer  
0                0                   0  
1                0                   0

In [10]:
df['rating'].value_counts()

4.5    3527
4.0    2853
3.5    1051
5.0     872
0.0     398
3.0     308
2.5      88
2.0      39
1.0      23
1.5       9
Name: rating, dtype: int64

In [11]:
df['category'].value_counts()

Perfume                      665
Moisturizers                 451
Face Serums                  384
Value & Gift Sets            378
Face Wash & Cleansers        247
                            ... 
Wellness                       1
High Tech Tools                1
Hair Styling & Treatments      1
Curls & Coils                  1
Lid Shadow Brush               1
Name: category, Length: 143, dtype: int64

In [12]:
df.describe().T

count          mean            std   min        25%  \
id                  9168.0  1.962952e+06  385971.360901  50.0  1819453.0   
rating              9168.0  3.990020e+00       1.007707   0.0        4.0   
number_of_reviews   9168.0  2.821392e+02     890.642028   0.0       10.0   
love                9168.0  1.627859e+04   42606.514471   0.0     1600.0   
price               9168.0  5.006324e+01      47.164989   2.0       24.0   
value_price         9168.0  5.182359e+01      49.459020   2.0       25.0   
online_only         9168.0  2.348386e-01       0.423921   0.0        0.0   
exclusive           9168.0  2.647251e-01       0.441211   0.0        0.0   
limited_edition     9168.0  9.184119e-02       0.288817   0.0        0.0   
limited_time_offer  9168.0  3.272251e-04       0.018087   0.0        0.0   

                          50%        75%        max  
id                  2072353.5  2230591.0  2359685.0  
rating                    4.0        4.5        5.0  
number_of_reviews        46.0      210.0    19000.0  
love                   4800.0    13800.0  1300000.0  
price                    35.0       59.0      549.0  
value_price              35.0       60.0      549.0  
online_only               0.0        0.0        1.0  
exclusive                 0.0        1.0        1.0  
limited_edition           0.0        0.0        1.0  
limited_time_offer        0.0        0.0        1.0

In [13]:
df.duplicated().sum()

0

2- Extraer los siguentes estadísiticos del conjunto de datos para varias de las variables disponibles:
- Media
- Desviación con respecto de la media
- Varianza
- Desviación típica
- Moda
- Mediana
- Media ponderada
- Robustez
- Coeficiente de variación


### Medias

---

In [14]:
# Media de precios
media_precio = df['price'].mean().round(2)
media_precio

50.06

In [15]:
# Media valor del producto
media_valor = df['value_price'].mean().round(2)
media_valor

51.82

In [16]:
# Media valoracion usuarios
media_rating = df['rating'].mean().round(2)
media_rating

3.99

El precio medio de los productos es 50,06€ y el valor medio 51,82€, son precios muy cercanos con lo que podemos concluir que no hay grandes descuentos en los productos, estan los precios muy ajustados a su valor.

Por otro lado la media de las valoracion es un 3,99 sobre 5 con lo cual podemos pensar que los productos en general tienen buenas valoraciones.

### Desviación con respecto a la media

---

In [17]:
# Precio
desv_media_precio = np.abs(df['price'] - media_precio)
desv_media_precio

0        15.94
1        15.94
2       129.94
3        69.94
4        21.94
         ...  
9163     44.06
9164     35.06
9165     34.06
9166      0.06
9167      0.06
Name: price, Length: 9168, dtype: float64

In [18]:
#Valor
desv_media_valor_precio = np.abs(df['value_price'] - media_valor)
desv_media_valor_precio

0        23.18
1        14.18
2       128.18
3        68.18
4        28.18
         ...  
9163     45.82
9164     24.82
9165     35.82
9166      1.82
9167      1.82
Name: value_price, Length: 9168, dtype: float64

In [19]:
#Valoracion
desv_media_rating= np.abs(df['rating'] - media_rating)
desv_media_rating

0       0.01
1       0.51
2       0.51
3       0.51
4       0.49
        ... 
9163    1.99
9164    3.99
9165    3.99
9166    1.01
9167    3.99
Name: rating, Length: 9168, dtype: float64

Con este calculo debido a la gran cantidad de datos no se observa claramente ninguna conclusión significativa. A simple vista podemos ver que hay productos que se separan mucho de la media de precios y valor y aparentemente en las valoraciones hay menos desviacion salvo en algunos productos no valorados.

### Varianza

---

In [20]:
# Precio
varianza_precio = df['price'].var()
varianza_precio

2224.5361947109777

In [21]:
# Valor
varianza_valor = df['value_price'].var()
varianza_valor

2446.194684800295

In [22]:
# Valoracion
varianza_rating = df['rating'].var()
varianza_rating

1.0154725424311082

Como la varianza del precio y valor es muy alta podemos suponer que hay una alta dispersión de los datos, es decir algún valor atípicos, productos con precios muy por encima o muy por debajo de la media, con lo cual podemos concluir que la media no es muy representativa en este caso.

Por otro lado la varianza de la valoración al ser muy cercana a 0 nos indica que los valores son mucho menos dispersos respecto a la media, y por tanto esta es más representativa.

### Desviación típica

---

In [23]:
# Precio
desvtipica_precio = df['price'].std()
desvtipica_precio

47.16498907782104

In [24]:
#Valor
desvtipica_valor = df['value_price'].std()
desvtipica_valor

49.45902025718155

In [25]:
# Valoracion
desvtipica_rating = df['rating'].std()
desvtipica_rating

1.0077065755621069

Podemos confirmar que la dispersión de los precios es de casi 50€, siendo muy significativa ya que es casi el mismo valor que la media, tanto en su precio como en el valor.

En cuanto a las valoraciones la desviación es de 1 punto, que podría considerarse alta ya que los datos van del 1 al 5, por lo que podemos considear que hay varios productos sin valorar.

### Moda

---

In [26]:
# Precio
moda_precio = df['price'].mode()[0]
moda_precio

25.0

In [27]:
# Valor
moda_valor = df['value_price'].mode()[0]
moda_valor

30.0

In [28]:
# rating
moda_rating = df['rating'].mode()[0]
moda_rating

4.5

Vemos que el precio más repetido son 25€ que se encuentra debajo de la media, con lo que concluimos que hay algún producto con precios muy elevados que descompensan la media, tanto en el precio como en el valor.

En cuanto a la valoración vemos que la más repetida es 4.5 puntos, al compararlo con la media podemos confirmar que hay muchos productos sin valorar que hacen disminuir la media a 3.99.

### Mediana

---

In [29]:
# Precio
mediana_precio = df['price'].median()
mediana_precio

35.0

In [30]:
# Valor
mediana_valor = df['value_price'].median()
mediana_valor

35.0

In [31]:
# rating
mediana_rating = df['rating'].median()
mediana_rating

4.0

La mediana, al estar más cerca de la moda que de la media nos confirma lo comentado anteriormente de productos con precios excesivamente altos con respecto al resto.

La mediana de las valoraciones sin embargo no dista tanto de la media con lo que podemos concluir que la media es bastante representativa

### Media ponderada

---

In [32]:
# Precio
w = np.random.randint(0,100, df.shape[0])

avg_precio = np.average(df["price"], weights=w).round()
avg_precio

50.0

In [33]:
# valor
w = np.random.randint(0,100, df.shape[0])

avg_valor = np.average(df["value_price"], weights=w).round()
avg_valor

52.0

In [34]:
# rating
w = np.random.randint(0,100, df.shape[0])

avg_rating = np.average(df["rating"], weights=w).round()
avg_rating

4.0

En este caso, como los pesos estan otorgados aleatoríamente la media ponderada tiene unos valores similares a la media aritmética. No nos aporta información ya que no consideramos que ningún producto tenga que tener un peso diferente a otro.

### Robustez

---

In [35]:
robustez = 1/df.size
print(robustez)

5.194049696667498e-06


Calculamos la robustez de la media siendo esta muy cercana a 0, con lo que el punto de ruptura de la media es muy bajo y por lo tanto poco robusta, se ve muy afectada por valores atipicos.

### Coeficiente de variacion

---

In [36]:
# Precio
coef_var_precio = desvtipica_precio / media_precio 
coef_var_precio

0.9421691785421702

In [37]:
# valor
coef_var_valor = desvtipica_valor / media_valor
coef_var_valor

0.9544388316708133

In [38]:
# rating
coef_var_rating = desvtipica_rating / media_rating 
coef_var_rating

0.25255803898799667

Al valorar el coeficiente de variación nos damos cuenta de que los datos de los precios y valores son muy heterogéneos  y sin embargo los ratings tiene valores mucho más homogéneos.